monty-31/src/x86_64_avx512/utils.rs

```diff
@@ -0,0 +1,221 @@
+use core::arch::x86_64::{self, __m512i};
+use core::mem::transmute;
+
+use crate::{MontyParameters, PackedMontyParameters, TwoAdicData};
+
+// Godbolt file showing that these all compile to the expected instructions. (Potentially plus a few memory ops):
+// https://godbolt.org/z/dvW7r1zjj
+
+/// Halve a vector of Monty31 field elements in canonical form.
+/// If the inputs are not in canonical form, the result is undefined.
+#[inline(always)]
+pub fn halve_avx512<MP: MontyParameters>(input: __m512i) -> __m512i {
+    /*
+        We want this to compile to:
+            vptestmd least_bit, val, ONE
+            vpsrld   t, val, 1
+            vpaddd   res, least_bit, t, maybe_half
+        throughput: 2 cyc/vec
+        latency: 4 cyc
+
+        Given an element val in [0, P), we want to compute val/2 mod P.
+        If val is even: val/2 mod P = val/2 = val >> 1.
+        If val is odd: val/2 mod P = (val + P)/2 = (val >> 1) + (P + 1)/2
+    */
+    unsafe {
+        // Safety: If this code got compiled then AVX2 intrinsics are available.
+        const ONE: __m512i = unsafe { transmute([1; 16]) };
```

Comment:
Make this more explicit:
```suggestion
        const ONE: __m512i = unsafe { transmute([1u32; 16]) };
```